staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/zz_generated.validations.go

```diff
@@ -0,0 +1,240 @@
+
+func Validate_DU1(in *DU1, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.DiscriminatedUnion(fldPath, in, dU1UnionMembership, in.D1)...)
+	// TypeMeta
+
+	// D1
+	errs = append(errs, Validate_D1(&in.D1, fldPath.Child("d1"))...)
+
+	return errs
+}
+
+var dU2UnionMembership = validate.NewDiscriminatedUnionMembership("d2", "m1", "m2")
+
+func Validate_DU2(in *DU2, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.DiscriminatedUnion(fldPath, in, dU2UnionMembership, in.D2, in.M1, in.M2)...)
+	// TypeMeta
+
+	// D2
+	errs = append(errs, Validate_D2(&in.D2, fldPath.Child("d2"))...)
+
+	// M1
+	if in.M1 != nil {
+		errs = append(errs, Validate_M1(in.M1, fldPath.Child("m1"))...)
+	}
+
+	// M2
+	if in.M2 != nil {
+		errs = append(errs, Validate_M2(in.M2, fldPath.Child("m2"))...)
+	}
+
+	// T1
+	if in.T1 != nil {
+		errs = append(errs, Validate_T1(in.T1, fldPath.Child("t1"))...)
+	}
+
+	return errs
+}
+
+var dU3UnionMembership = validate.NewDiscriminatedUnionMembership("d2", [2]string{"m1", "CustomM1"}, [2]string{"m2", "CustomM2"})
+
+func Validate_DU3(in *DU3, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.DiscriminatedUnion(fldPath, in, dU3UnionMembership, in.D3, in.M1, in.M2)...)
+	// TypeMeta
+
+	// D3
+	errs = append(errs, Validate_D3(&in.D3, fldPath.Child("d2"))...)
+
+	// M1
+	if in.M1 != nil {
+		errs = append(errs, Validate_M1(in.M1, fldPath.Child("m1"))...)
+	}
+
+	// M2
+	if in.M2 != nil {
+		errs = append(errs, Validate_M2(in.M2, fldPath.Child("m2"))...)
+	}
+
+	// T1
+	if in.T1 != nil {
+		errs = append(errs, Validate_T1(in.T1, fldPath.Child("t1"))...)
+	}
+
+	return errs
+}
+
+func Validate_M1(in *M1, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.FixedResult(fldPath, in, true, "type M1")...)
+	// S
+	errs = append(errs, validate.FixedResult(fldPath.Child("s"), in.S, true, "field M1.S")...)
+
+	return errs
+}
+
+func Validate_M2(in *M2, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.FixedResult(fldPath, in, true, "type M2")...)
+	// S
+	errs = append(errs, validate.FixedResult(fldPath.Child("s"), in.S, true, "field M2.S")...)
+
+	return errs
+}
+
+func Validate_T1(in *T1, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.FixedResult(fldPath, in, true, "type T1")...)
+	// TypeMeta
+
+	// LS
+	errs = append(errs, validate.FixedResult(fldPath.Child("ls"), in.LS, true, "field T1.LS")...)
+	for i, val := range in.LS {
+		errs = append(errs, validate.FixedResult(fldPath.Index(i), val, true, "field T1.LS[*]")...)
+		errs = append(errs, validate.Required(fldPath.Index(i), val)...)
+	}
+
+	return errs
+}
+
+var u1UnionMembership = validate.NewUnionMembership("m1", "m2")
+
+func Validate_U1(in *U1, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.Union(fldPath, in, u1UnionMembership, in.M1, in.M2)...)
```

Comment:
So this reports an error on the containing struct, which really precludes having 2 unions in a struct, right?

Shouldn't this be fatal wrt the union members?  E.g. 

```
if e := validate.Union(fldPath, in, u1UnionMembership, in.M1, in.M2); len(e) != 0 {
    errs = append(errs, e...)
} else {
    // validate M1
    // validate M2
}
```

I think the func() thing might make this easier?

Reply:
Thinking more about this, it's tricky.  How would we express this?

One idea before I forget it:
* add an optional "Dependants()" method to GenFunction.  It returns a list of fields whose validation can only proceed if *this* validation succeeds  
* union validator returns the union members from Dependants()
* code generates something like:

```
Validate_U1(...) {
    failUnion<hash> := false
    if e := validate.Union(...); len(e) != 0 {
        errs = append(...
        failUnion<hash> = true
    }

    // M1
    if !failUnion<hash> {
        // other M1 validation
    }
```

Reply:
Trying to find other ways to approach this.  Thinking out loud.

Maybe make less scaffolding for generators.  Just pass them a SnippetWriter and maybe an Args, and let them emit code with a given assumed environment.

E.g.  Instead of calling `v.ExtractValidations()` and then iterating, just call `v.Render(sw)`.  Each validator can assume:
* `obj` is the object in question
* `fldPath` is the current field path
* `errs` is where to append results
* they can `return` at will, and bypass any further validation

The `required` validator would emit the `if e:= ...; len(e) != 0` logic, or maybe we keep the flags stuff to make it a bit easier.  Same for nil-checks.

The `union` validator... this doesn't help, does it?  We still need a file-scope variable and we still need to influence OTHER field validations.  I'll keep thinking about it.  I am pretty much out of time to review further, but I may yet bring laptop on vacation because this is fun.

Reply:
Oh boy, yes. This is a big gap. Looking more closely.

Reply:
I'll hold on this until dependent fields is sorted out.